### ISSUE TITLE
About the learning rate schedule and text-text fusion

### ISSUE THREAD

**Issue by dandelin** [Wangt-CN/MTFN-RR-PyTorch-Code#1]
Thanks for sharing the code, nice work!

I have a question on the learning rate scheduling of MTFN.
In the paper, you've wrote that the network was trained for 50 epochs.
However, it seems the code's configuration was tuned to 160 epochs training.
Plus, it seems the code doesn't contain the text-text fusion.

Could you provide more details on the scheduling and recipe for training the text-text fusion (such as "How to sample the captions to make triplet")?
Thanks again :)

**Comment by Wangt-CN** [comment]
Hi, Thanks for your try.
1, What dataset are you use? MSCOCO or Flickr30k. However, the epoches about 160 is very strange. For my own training, 50 to 70 epoches can get the best result, whatever on MSCOCO or Flickr.
Can you provide more details about your training?

2, Yes, the code now didn't provide text fusion so maybe you can't directly use it. And I'm still on preparing this part. However, as we wrote in the paper, this structure is same as image-text.

**Comment by dandelin** [comment]
Thanks for reply!

1. I didn't choose the number 160, it is written in your config file. (https://github.com/Wangt-CN/MTFN-RR-PyTorch-Code/blob/master/option/FusionNoattn_baseline.yaml)
Plus, the learning rate scheduling is bit different from the paper (https://github.com/Wangt-CN/MTFN-RR-PyTorch-Code/blob/master/train.py#L172-L183), which multiplies 0.5 for every 10 epoch.

2. If we use precomp dataset, there is a possibility that the minibatch does not contain any positive text pair (for instance, all 128 examples are different images).
For my understanding, the training of Image-Text fusion and Text-Text fusion occurs simultaneously.
So I am wondering what sampling scheme guarantees the positive pair to build a triplet (Eq. 7 in the paper).

**Comment by Wangt-CN** [comment]
Thanks your point out.
1, Yes, sorry, this version is an afterward fine-tune version and please let me check again my code. Moreover, the model still needn't such 160 epoches to converge, it's just a setting to see if the performance can be better. You can try more settings about the learning rate and batchszie, etc.

2, The training of image-text fusino and text-text fusion is not at the same time. The reason is as you said, we can not promise the good data sample. So the text-text fusion is another training procedure but data loded in is the text data. And it need not so much time as image-text to converge. However I need more time to prepare this part of code.
Moreover, for the last step "rerank", you can also use it without text-text fusion result and it can still boost the result of I2T. The text-text fusion result is used to boost T2I

**Comment by dandelin** [comment]
Thanks for quick reply.
It really helps me to understanding your work :)

1. Yes, I also found that the results converges after 50 epochs for both datasets.

2. Now I understand the term _optimized independently_ in your paper 👍 .
But the training scheme is not clearly written in the paper.
In my guess, the optimization of two losses could be either one of these:
  1) alternating every iteration (optimize L(Ip, Tp) for a image-text minibatch and optimize L(Tp, Tq) for a text-text minibatch).
  2) alternating every epoch (optimize L(Ip, Tp) for a entire multimodal training set and optimize L(Tp, Tq) for a entire text-only training set).
  3) or finetune a model trained with L(Ip, Tp) using L(Tp, Tq).

Correct me if the scheme is not one of those, it will greatly help me out 😄 .

**Comment by Wangt-CN** [comment]
Hi, 
Actually is the 3) scheme. Because the two model structure is same so we can initialize the text-text model's parameters with the image-text. And in our experiment it's very essential for quikly convergence.
However, I think the 2) scheme you say is a very good scheme. And if it can work, the modal will be an entire end-to-end model. I can't have a try because I may not have more bandwidth these days, however if you have any interests on it, very welcome your trying and report or pr. :)

Finally, welcome for using and  recommending our code, especially the `rerank.py`  scheme. I really think it's a very good-use script with a few seconds time consuming to boost the iamge-text matching performance.

**Comment by dandelin** [comment]
Yes, I was impressed by the re-ranking method was used other than the task of landmark retrieval.
Since it can mixed into almost every cross-modal retrieval model, I will definitely try it in my further experiments. :)

Please let me know when you complete organizing the code for text-text fusion.
I'll close this issue till then.
Thank you 👍
